File: apps/api/app/schemas/virtual.py
"""虚拟交易 Pydantic 契约 · 0008 v2 三独立子账户。

所有金额字段原币种 Decimal · 不做 CNY 折算。
"""

from __future__ import annotations

from decimal import Decimal
from typing import Literal

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field

from app.models.virtual import Currency, OrderSide, OrderStatus, OrderType, PositionSide
from app.schemas.market import Market

# ===== Account =====


class AccountActivateIn(BaseModel):
    """PUT /accounts/{market} 入参 · 激活或重置子账户。"""

    model_config = ConfigDict(extra="forbid")

    initial_capital: Decimal = Field(
        gt=0,
        le=Decimal("999999999"),
        description="初始资金(原币种 · 上限 9.99 亿,避免误填天文数字)",
    )


class AccountResponse(BaseModel):
    """子账户响应 · 单市场。"""

    model_config = ConfigDict(extra="forbid", frozen=True, from_attributes=True)

    id: int
    market: Market
    currency: Currency
    initial_capital: Decimal
    cash_balance: Decimal
    realized_pnl: Decimal
    activated_at: AwareDatetime


# ===== Position =====


class PositionResponse(BaseModel):
    """持仓响应(活仓 + 历史共用)· 含实时浮盈(activate 时)。"""

    model_config = ConfigDict(extra="forbid", frozen=True, from_attributes=True)

    id: int
    symbol: str
    market: Market
    position_side: PositionSide
    quantity: Decimal
    avg_entry_price: Decimal
    realized_pnl: Decimal | None
    opened_at: AwareDatetime
    closed_at: AwareDatetime | None  # NULL = 活仓


class PositionWithQuoteResponse(BaseModel):
    """持仓 + 实时价 + 浮盈(给 /portfolio 用)。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    id: int
    symbol: str
    market: Market
    position_side: PositionSide
    quantity: Decimal
    avg_entry_price: Decimal
    current_price: Decimal | None
    unrealized_pnl: Decimal | None
    value: Decimal | None  # long: quantity × current_price · short: 担保 + 浮盈


# ===== Portfolio =====


class AccountSummaryResponse(BaseModel):
    """单市场账户全貌:account + 活仓 + 实时估值。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    account_id: int
    market: Market
    currency: Currency
    initial_capital: Decimal
    cash_balance: Decimal
    realized_pnl: Decimal
    positions: list[PositionWithQuoteResponse]
    positions_value: Decimal
    total_equity: Decimal


# ===== Order =====


class OrderPlaceIn(BaseModel):
    """POST /orders 入参 · 市价单。"""

    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(min_length=1, max_length=64)
    market: Market
    side: OrderSide
    # 3.4 · 默认 LONG(做多 · A股/加密/美股做多)· SHORT 仅美股卖空(SELL=开空 / BUY=平空)
    position_side: PositionSide = PositionSide.LONG
    quantity: Decimal = Field(gt=0)


class OrderResponse(BaseModel):
    """订单流水响应(filled / rejected)。"""

    model_config = ConfigDict(extra="forbid", frozen=True, from_attributes=True)

    id: int | None  # 未激活市场拒单时为 None(临时对象,未持久化)
    account_id: int | None
    symbol: str
    market: Market
    side: OrderSide
    order_type: OrderType
    position_side: PositionSide
    quantity: Decimal
    price: Decimal | None
    notional: Decimal | None
    commission: Decimal | None
    slippage_cost: Decimal | None
    realized_pnl: Decimal | None
    status: OrderStatus
    reject_reason: str | None
    placed_at: AwareDatetime | None
    filled_at: AwareDatetime | None


class HkBoardLotResponse(BaseModel):
    """港股每手股数(board lot)· 前端下单按手取整 / UI 提示用 · 只读。

    lot 来自 hk_pool(策展 18 只 · 已逐一核港交所官方)· 不在池内 → 404。
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    symbol: str
    board_lot: int


# ===== AI 模拟下单(0036 批次甲)=====


class AiOrderRequest(BaseModel):
    """POST /virtual/ai-order 入参 · AI 建议一键模拟下单。

    ★ 红线:只走现有虚拟撮合引擎(execute → place_market_order / route_open/close_perp),
      绝不接真实交易。direction 来自 AI 决策卡 actionable 建议(hold/close 不经此端点)。
    仓位不传:由 execute 内部按用户 BotOrderPreset 派生(拍板③)。
    """

    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(min_length=1, max_length=64)
    market: Market
    # 仅 4 个可下单方向 · cn/us 用 buy/sell · crypto 用 open_long/open_short(合约)
    direction: Literal["buy", "sell", "open_long", "open_short"]


class AiOrderResponse(BaseModel):
    """AI 模拟下单结果 · 与手动下单同一虚拟引擎成交 · 标 source=ai_signal。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    filled: bool
    title: str
    detail: str
    source: str = "ai_signal"


# ===== Equity curve =====


class EquitySnapshotResponse(BaseModel):
    """权益曲线点。"""

    model_config = ConfigDict(extra="forbid", frozen=True, from_attributes=True)

    cash: Decimal
    positions_value: Decimal
    equity: Decimal
    realized_pnl_cumulative: Decimal
    snapshot_at: AwareDatetime


class EquityCurvesResponse(BaseModel):
    """多市场曲线 · 按 market 分组。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    curves: dict[str, list[EquitySnapshotResponse]]

File: apps/web/lib/api/virtual.ts
/**
 * 虚拟交易 API client · 0008 v2 三独立子账户。
 *
 * 8 个 endpoint 全部需要 Bearer JWT(从 NextAuth session.accessToken 取)。
 * 金额字段全部 string(后端 Decimal 序列化为 string,避免 JS 浮点精度损失)。
 */

import type { Market } from '@midas/shared'

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000'

export type Currency = 'CNY' | 'USD' | 'USDT'
export type OrderSide = 'buy' | 'sell'
export type OrderType = 'market'
export type OrderStatus = 'filled' | 'rejected'
// 3.4 · 持仓方向 · long=做多(A股/加密/美股做多)· short=美股卖空(SELL 开空 / BUY 平空)
export type PositionSide = 'long' | 'short'

export interface VirtualAccount {
  id: number
  market: Market
  currency: Currency
  initial_capital: string
  cash_balance: string
  realized_pnl: string
  activated_at: string
}

export interface PositionView {
  id: number
  symbol: string
  market: Market
  position_side: PositionSide
  quantity: string
  avg_entry_price: string
  current_price: string | null
  unrealized_pnl: string | null
  value: string | null
}

export interface AccountSummary {
  account_id: number
  market: Market
  currency: Currency
  initial_capital: string
  cash_balance: string
  realized_pnl: string
  positions: PositionView[]
  positions_value: string
  total_equity: string
}

export interface Position {
  id: number
  symbol: string
  market: Market
  position_side: PositionSide
  quantity: string
  avg_entry_price: string
  realized_pnl: string | null
  opened_at: string
  closed_at: string | null
}

export interface VirtualOrder {
  id: number | null
  account_id: number | null
  symbol: string
  market: Market
  side: OrderSide
  position_side: PositionSide
  order_type: OrderType
  quantity: string
  price: string | null
  notional: string | null
  commission: string | null
  slippage_cost: string | null
  realized_pnl: string | null
  status: OrderStatus
  reject_reason: string | null
  placed_at: string | null
  filled_at: string | null
}

export interface EquitySnapshot {
  cash: string
  positions_value: string
  equity: string
  realized_pnl_cumulative: string
  snapshot_at: string
}

export interface EquityCurves {
  curves: Record<string, EquitySnapshot[]>
}

export class VirtualApiError extends Error {
  constructor(
    public readonly status: number,
    public readonly detail: string,
  ) {
    super(`VirtualApi ${status}: ${detail}`)
    this.name = 'VirtualApiError'
  }
}

async function readDetail(resp: Response): Promise<string> {
  try {
    const body = (await resp.json()) as { detail?: unknown }
    return typeof body.detail === 'string' ? body.detail : `HTTP ${resp.status}`
  } catch {
    return `HTTP ${resp.status}`
  }
}

function authHeaders(token: string): HeadersInit {
  return {
    Authorization: `Bearer ${token}`,
    'Content-Type': 'application/json',
  }
}

// ===== Accounts =====

export async function fetchAccounts(
  token: string, signal?: AbortSignal,
): Promise<VirtualAccount[]> {
  const r = await fetch(`${API_BASE}/api/v1/virtual/accounts`, {
    headers: authHeaders(token), signal,
  })
  if (!r.ok) throw new VirtualApiError(r.status, await readDetail(r))
  return (await r.json()) as VirtualAccount[]
}

export async function fetchAccount(
  token: string, market: Market, signal?: AbortSignal,
): Promise<VirtualAccount | null> {
  const r = await fetch(`${API_BASE}/api/v1/virtual/accounts/${market}`, {
    headers: authHeaders(token), signal,
  })
  if (r.status === 404) return null
  if (!r.ok) throw new VirtualApiError(r.status, await readDetail(r))
  return (await r.json()) as VirtualAccount
}

export async function activateOrResetAccount(
  token: string, market: Market, initialCapital: string,
): Promise<VirtualAccount> {
  const r = await fetch(`${API_BASE}/api/v1/virtual/accounts/${market}`, {
    method: 'PUT',
    headers: authHeaders(token),
    body: JSON.stringify({ initial_capital: initialCapital }),
  })
  if (!r.ok) throw new VirtualApiError(r.status, await readDetail(r))
  return (await r.json()) as VirtualAccount
}

// ===== Portfolio =====

export async function fetchPortfolio(
  token: string, signal?: AbortSignal,
): Promise<AccountSummary[]> {
  const r = await fetch(`${API_BASE}/api/v1/virtual/portfolio`, {
    headers: authHeaders(token), signal,
  })
  if (!r.ok) throw new VirtualApiError(r.status, await readDetail(r))
  return (await r.json()) as AccountSummary[]
}

// ===== Orders =====

export interface PlaceOrderInput {
  symbol: string
  market: Market
  side: OrderSide
  quantity: string
  // 3.4 · 不传 = long(后端默认 · A股/加密/美股做多原行为不变)· short 仅美股卖空
  position_side?: PositionSide
}

export async function placeOrder(
  token: string, input: PlaceOrderInput,
): Promise<VirtualOrder> {
  const r = await fetch(`${API_BASE}/api/v1/virtual/orders`, {
    method: 'POST',
    headers: authHeaders(token),
    body: JSON.stringify(input),
  })
  if (!r.ok) throw new VirtualApiError(r.status, await readDetail(r))
  return (await r.json()) as VirtualOrder
}

export async function fetchOrders(
  token: string,
  opts: { market?: Market; limit?: number; beforeId?: number } = {},
  signal?: AbortSignal,
): Promise<VirtualOrder[]> {
  const params = new URLSearchParams()
  if (opts.market) params.set('market', opts.market)
  if (opts.limit) params.set('limit', String(opts.limit))
  if (opts.beforeId) params.set('before_id', String(opts.beforeId))
  const r = await fetch(
    `${API_BASE}/api/v1/virtual/orders?${params.toString()}`,
    { headers: authHeaders(token), signal },
  )
  if (!r.ok) throw new VirtualApiError(r.status, await readDetail(r))
  return (await r.json()) as VirtualOrder[]
}

// ===== Positions =====

export async function fetchPositions(
  token: string,
  opts: { includeClosed?: boolean; market?: Market } = {},
  signal?: AbortSignal,
): Promise<Position[]> {
  const params = new URLSearchParams()
  if (opts.includeClosed) params.set('include_closed', 'true')
  if (opts.market) params.set('market', opts.market)
  const r = await fetch(
    `${API_BASE}/api/v1/virtual/positions?${params.toString()}`,
    { headers: authHeaders(token), signal },
  )
  if (!r.ok) throw new VirtualApiError(r.status, await readDetail(r))
  return (await r.json()) as Position[]
}

// ===== Equity curves =====

export async function fetchEquityCurves(
  token: string, days = 30, signal?: AbortSignal,
): Promise<EquityCurves> {
  const r = await fetch(
    `${API_BASE}/api/v1/virtual/equity-curves?days=${days}`,
    { headers: authHeaders(token), signal },
  )
  if (!r.ok) throw new VirtualApiError(r.status, await readDetail(r))
  return (await r.json()) as EquityCurves
}
